### ISSUE TITLE
[css-scroll-snap-2] How should competing scroll-start-targets be resolved?

### ISSUE THREAD

**Issue by DavMila** [w3c/csswg-drafts#10774]
The scroll snap 2 [spec](https://drafts.csswg.org/css-scroll-snap-2/#scroll-start-target) says that when there are multiple elements that could be scroll-start-targets for a scroll container "user-agents should select the one which comes first in [tree order](https://dom.spec.whatwg.org/#concept-tree-order)".
This feels like a simple solution that meets the needs of scroll-start-target use cases. It's perhaps also a natural solution as it's somewhat similar to the css-scroll-snap spec's precedent of falling back on tree order when [selecting between multiple aligned snap areas](https://drafts.csswg.org/css-scroll-snap/#multiple-aligned-snap-areas).

Alternatively, we could compute the intersection of all relevant scroll-start-targets ([suggested](https://github.com/w3c/csswg-drafts/issues/8832#issuecomment-1548232540) by @flackr ). We could compute the relevant [scroll-into-view positions](https://drafts.csswg.org/cssom-view-1/#determine-the-scroll-into-view-position) (or some other method) of all the scroll-start-targets and then use the average along each axis. Since this could result in none of the scroll-start-targets ending up in the scrollport, we would handle this case by falling back to the first in tree order.

I propose we stick with what is currently spec'd as the use cases for scroll-start-target that I can think of (e.g. carousels, drop-down menu starting at an arbitrary item) really only intend scroll to a single target, not some average of multiple targets.

**Comment by flackr** [comment]
I thought of a possible option that would support a lot of the 2d cases fairly simply. We could scroll to every specified item in order, then if you combine this with `nearest` behavior it will try to scroll to a position that includes all of the specified elements. Elements that are wider than the viewport (e.g. scrolling to a row in a table) won't move the horizontal position as its already fully in view and thus implicitly supports the scrolling to a specific row and column use case.

I also wonder if we should consider re-using `scroll-snap-align` for this. We already use `scroll-margin` and `scroll-padding` to affect scrolling behaviors beyond snapping, it seems reasonable that we may want a `scroll-align` which affects the alignment used for anchor scrolls or other scrolls even when not using scroll snapping.

**Comment by argyleink** [comment]
great idea to leverage `snap-align`, that way the `scroll-start-target` when scrolled to can be positioned optimally 👍🏻

**Comment by flackr** [comment]
Some use cases to consider where multiple scroll start targets might be specified for a single scrolling container could be:
1. A long list where the author specifies all checked items to start scrolled into view (e.g. start scrolled to the items you have selected so you can quickly review them and unselect them if desired).
2. A navigation map view where the start, end, and possibly intermediate stops on your journey are scroll start targets so that the map starts scrolled to the planned route.
3. A table where a particular row and column start scrolled into view.
4. A table where multiple arbitrary cells are requested to start scrolled into view (e.g. maybe they point to key values the viewer is expected to be interested in).

**Comment by flackr** [comment]
To address the use cases and resolve this issue as well as #8832, I propose the following:

1. scroll-start-target does not have per-axis values, i.e. no change to [the spec](https://drafts.csswg.org/css-scroll-snap-2/#scroll-start-target-propdef) as it is currently written.
2. When multiple elements are requested to be scroll start targets, determine the scroll position by scrolling to each one in reverse-dom order.
3. Define `scroll-align`, a generalized version of `scroll-snap-align` that works outside of snap containers and possibly in snap containers without establishing a snap point. This former is [currently a MAY in the spec](https://drafts.csswg.org/css-scroll-snap-1/#choosing):
   > If a page is navigated to a fragment that defines a target element (e.g. one that would be matched by [:target](https://drafts.csswg.org/selectors-4/#target-pseudo), or the target of [scrollIntoView()](https://drafts.csswg.org/cssom-view-1/#dom-element-scrollintoview)) ... the user agent must [snap](https://drafts.csswg.org/css-scroll-snap-1/#scroll-snap) to one of that element’s snap positions if its nearest [scroll container](https://drafts.csswg.org/css-overflow-3/#scroll-container) is a [scroll snap container](https://drafts.csswg.org/css-scroll-snap-1/#scroll-snap-container). The user agent **may** also do this even when the scroll container has [scroll-snap-type: none](https://drafts.csswg.org/css-scroll-snap-1/#propdef-scroll-snap-type).

   This property should also accept `nearest` values to allow specifying all of the [ScrollLogicalPosition](https://drafts.csswg.org/cssom-view/#enumdef-scrolllogicalposition) values of the [scrollIntoView](https://drafts.csswg.org/cssom-view/#dom-element-scrollintoview) API. It may make sense for scroll snapping to support `nearest` as well to support making items that snap such that they're not partially in view but don't have a particular affinity for aligning only on one side of the scrollport.

The combination of `nearest` values in point 3 and attempting to scroll to multiple items in point 2 would provide a reasonable best-effort attempt to scroll all of the requested items into view while still ensuring that the first one (the current specified behavior) is in view at the end.

**Comment by DavMila** [comment]
>    This property should also accept `nearest` values to allow specifying all of the [ScrollLogicalPosition](https://drafts.csswg.org/cssom-view/#enumdef-scrolllogicalposition) values of the [scrollIntoView](https://drafts.csswg.org/cssom-view/#dom-element-scrollintoview) API. It may make sense for scroll snapping to support `nearest` as well to support making items that snap such that they're not partially in view but don't have a particular affinity for aligning only on one side of the scrollport.

This proposal makes sense to me overall but I could use some help clarifying the concept of "items that snap such that they're not partially in view but don't have a particular affinity for aligning only on one side of the scrollport." Is this similar to scrollIntoView's concept of nearest where if the element is already partially in view, no scrolling happens?

**Comment by flackr** [comment]
> I could use some help clarifying the concept of "items that snap such that they're not partially in view but don't have a particular affinity for aligning only on one side of the scrollport." Is this similar to scrollIntoView's concept of nearest where if the element is already partially in view, no scrolling happens?

Yes, that is exactly right. `scroll-snap-align: nearest` would try to snap the element to the start and end alignment so that the scroll does not rest in a position with the element partially clipped by the scrolling container.

**Comment by DavMila** [comment]
Okay thanks, I think your proposal would be a great way to solve this. Shall we agenda+ this issue?
